quinn-udp/src/unix.rs

```diff
@@ -833,6 +844,118 @@ mod gso {
     pub(crate) fn set_segment_size(encoder: &mut cmsg::Encoder<libc::msghdr>, segment_size: u16) {
         encoder.push(libc::SOL_UDP, UDP_SEGMENT, segment_size);
     }
+
+    fn supported_by_current_kernel() -> bool {
+        let kernel_version_string = match kernel_version_string() {
+            Ok(kernel_version_string) => kernel_version_string,
+            Err(_errno) => {
+                crate::log::warn!("GSO disabled: uname returned {_errno}");
+                return false;
+            }
+        };
+
+        let kernel_version = match KernelVersion::from_str(&kernel_version_string) {
+            Ok(kernel_version) => kernel_version,
+            Err(_reason) => {
+                crate::log::warn!("GSO disabled: {_reason}");
+                return false;
+            }
+        };
+
+        if kernel_version < SUPPORTED_SINCE {
+            crate::log::info!("GSO disabled: kernel too old ({kernel_version_string}); need 4.18+",);
+            return false;
+        }
+
+        true
+    }
+
+    fn kernel_version_string() -> Result<String, libc::c_int> {
+        let mut n = unsafe { std::mem::zeroed() };
+        let r = unsafe { libc::uname(&mut n) };
+        if r != 0 {
+            return Err(r);
+        }
+        Ok(unsafe {
+            std::ffi::CStr::from_ptr(n.release[..].as_ptr())
+                .to_string_lossy()
+                .into_owned()
+        })
+    }
+
+    // https://www.linfo.org/kernel_version_numbering.html
+    #[derive(Eq, PartialEq, Ord, PartialOrd, Debug)]
+    struct KernelVersion {
+        version: u8,
+        major_revision: u8,
+    }
+
+    impl KernelVersion {
+        const fn new(version: u8, major_revision: u8) -> Self {
+            Self {
+                version,
+                major_revision,
+            }
+        }
+
+        fn from_str(release: &str) -> Result<Self, String> {
```

Comment:
Nit: suggest returning `Option<Self>` from this.

Reply:
That is what we originally had and I suggested returning an error to consolidate the places where we log what went wrong :)

Reply:
I guess as long as we print the original string returned from `uname` as part of the WARN log, we can always reproduce it if this gets hit in the wild.